lasagne/tests/layers/test_special.py

```diff
@@ -400,19 +498,48 @@ def test_transform_downsample():
         assert x_out2.shape == x_out.shape
         np.testing.assert_allclose(x_out2, x_out, rtol=1e-5, atol=1e-5)
 
-
-def test_transform_identity():
-    from lasagne.layers import InputLayer, TransformerLayer
-    from lasagne.utils import floatX
-    from theano.tensor import constant
-    batchsize = 10
-    l_in = InputLayer((batchsize, 3, 28, 28))
-    l_loc = InputLayer((batchsize, 6))
-    layer = TransformerLayer(l_in, l_loc)
-    inputs = floatX(np.arange(np.prod(l_in.shape)).reshape(l_in.shape))
-    thetas = floatX(np.tile([1, 0, 0, 0, 1, 0], (batchsize, 1)))
-    outputs = layer.get_output_for([constant(inputs), constant(thetas)]).eval()
-    np.testing.assert_allclose(inputs, outputs, rtol=1e-6)
+    def test_transform_thin_plate_spline_identity(self):
+        from lasagne.layers import InputLayer, TPSTransformerLayer
+        from lasagne.utils import floatX
+        from theano.tensor import constant
+        batchsize = 5
+        num_control_points = 16
+        dest_offset = np.zeros(shape=(batchsize, 2*num_control_points))
+        l_in = InputLayer((batchsize, 3, 28, 28))
+        l_loc = InputLayer((batchsize, 2*num_control_points))
+        layer = TPSTransformerLayer(
+                l_in, l_loc, control_points=num_control_points
+        )
+        inputs = floatX(np.arange(np.prod(l_in.shape)).reshape(l_in.shape))
+        outputs = layer.get_output_for([constant(inputs),
+                                        constant(dest_offset)]).eval()
+        np.testing.assert_allclose(inputs, outputs, atol=1e-5)
+
+    def test_transform_thin_plate_spline_shift(self):
+        from lasagne.layers import InputLayer, TPSTransformerLayer
+        from theano.tensor import constant
+        batchsize = 5
+        num_control_points = 16
+        dest_offset = np.ones(shape=(batchsize, 2*num_control_points))
+        l_in = InputLayer((batchsize, 3, 28, 28))
+        l_loc = InputLayer((batchsize, 2*num_control_points))
+        layer = TPSTransformerLayer(
+                l_in, l_loc, control_points=num_control_points
+        )
+        image = np.zeros(shape=(28, 28))
+        image[[0, -1], :] = 1
+        image[:, [0, -1]] = 1
+        inputs = np.tile(image, (batchsize, 3, 1, 1))
+        shifted_input = np.ones(shape=(28, 28))
+        shifted_input[:13, :13] = 0
+        shifted_input[13, :13] = 0.50000271
+        shifted_input[:13, 13] = 0.50000271
+        shifted_input[13, 13] = 0.75000271
```

Comment:
Yep, I'm pretty sure (though not 100% positive). I think it's due to the some minor rounding in the interpolation. I know for certain that the transformed points for the new grid are correct though because I've checked them against another [library](https://github.com/olt/thinplatespline).